### PR TITLE
Prometheus mesos service discovery

### PR DESCRIPTION
**Pull request by philipnrmn**
This PR adds a mesos task service discovery mechanism to the Prometheus plugin. 

Tasks may expose Prometheus format metrics for collection on nominated ports. The ports in question are discovered by Telegraf via Mesos state. Collection will be attempted throughout the life of the task. Failure to collect metrics will be logged, but has no other impact.

To allow the metrics exposed on any given port (or multiple ports), tasks may EITHER:

- Label the port with `DCOS_METRICS_FORMAT=prometheus`

OR

- Label the task with `DCOS_METRICS_FORMAT=prometheus`
- Label the task with the index of the metrics port eg `DCOS_METRICS_PORT=0`

### CODE REVIEW

File: plugins/inputs/prometheus/prometheus.go [hunk 1]
```diff
@@ -237,8 +287,151 @@ func (p *Prometheus) gatherURL(u URLAndAddress, acc telegraf.Accumulator) error
...
+		return r, nil
+	} else {
+		return r, fmt.Errorf("processResponse expected type %q, got %q", t, r.GetType())
+	}
+}
+
+// getMesosTaskPrometheusURLs converts a list of tasks to a list of Prometheus
+// URLs to scrape
+func getMesosTaskPrometheusURLs(tasks *agent.Response_GetTasks) []*url.URL {
+	results := []*url.URL{}
```
Comment (janisz) [inline plugins/inputs/prometheus/prometheus.go]: Why slice of pointers?

File: plugins/inputs/prometheus/mock_mesos_server.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"path/filepath"
+	"testing"
+)
+
+// raw protobuf request types:
```
Comment (janisz) [inline plugins/inputs/prometheus/mock_mesos_server.go:22]: Can you add a link to definitions of this values?

File: plugins/inputs/prometheus/prometheus.go (hunk 1 above)
Comment (philipnrmn) [reply plugins/inputs/prometheus/prometheus.go]: I moved the processing into the parent method to avoid this.

**Review by branden — APPROVED** [review]
Looks great. Would this be worth contributing upstream?

**Comment by philipnrmn** [comment]
Thanks! Yes I'd like to contribute it upstream too.

**Pull request merged**